block-producer/src/batch_builder/mod.rs

```diff
@@ -0,0 +1,205 @@
+    /// corresponding new hash
+    pub fn updated_accounts(&self) -> impl Iterator<Item = (AccountId, Digest)> + '_ {
+        self.txs.iter().map(|tx| (tx.account_id(), tx.final_account_hash()))
+    }
+
+    /// Returns the script root of all consumed notes
+    pub fn consumed_notes_script_roots(&self) -> impl Iterator<Item = Digest> + '_ {
+        let mut script_roots: Vec<Digest> = self
+            .txs
+            .iter()
+            .flat_map(|tx| tx.consumed_notes())
+            .map(|consumed_note| consumed_note.script_root())
+            .collect();
+
+        script_roots.sort();
+
+        // Removes duplicates in consecutive items
+        script_roots.into_iter().dedup()
+    }
+
+    /// Returns the nullifier of all consumed notes
+    pub fn consumed_notes_nullifiers(&self) -> impl Iterator<Item = Digest> + '_ {
+        self.txs
+            .iter()
+            .flat_map(|tx| tx.consumed_notes())
+            .map(|consumed_note| consumed_note.nullifier())
+    }
+
+    /// Returns the hash of created notes
+    pub fn created_notes_hashes(&self) -> impl Iterator<Item = Digest> + '_ {
+        self.txs
+            .iter()
+            .flat_map(|tx| tx.created_notes())
+            .map(|note_envelope| note_envelope.note_hash())
+    }
+}
+
+// BATCH BUILDER
+// ================================================================================================
+
+#[async_trait]
+pub trait BatchBuilder: Send + Sync + 'static {
+    // TODO: Make concrete `AddBatches` Error?
+    type AddBatchesError: Debug;
+
+    async fn add_tx_groups(
+        &self,
+        tx_groups: Vec<Vec<SharedProvenTx>>,
+    ) -> Result<(), Self::AddBatchesError>;
+}
+
+pub struct DefaultBatchBuilderOptions {
+    /// The frequency at which blocks are created
+    pub block_frequency: Duration,
+
+    /// Maximum number of batches in any given block
+    pub max_batches_per_block: usize,
+}
+
+pub struct DefaultBatchBuilder<BB>
+where
+    BB: BlockBuilder,
+{
+    /// Batches ready to be included in a block
+    ready_batches: SharedRwVec<Arc<TransactionBatch>>,
+
+    block_builder: Arc<BB>,
+
+    options: DefaultBatchBuilderOptions,
+}
+
+impl<BB> DefaultBatchBuilder<BB>
+where
+    BB: BlockBuilder,
+{
+    pub fn new(
+        block_builder: Arc<BB>,
+        options: DefaultBatchBuilderOptions,
+    ) -> Self {
+        Self {
+            ready_batches: Arc::new(RwLock::new(Vec::new())),
+            block_builder,
+            options,
+        }
+    }
+
+    pub async fn run(self) {
+        let mut interval = time::interval(self.options.block_frequency);
+
+        loop {
+            interval.tick().await;
+            self.try_send_batches().await;
+        }
+    }
+
+    /// Note that we call `add_batches()` regardless of whether the `ready_batches` queue is empty.
+    /// A call to an empty `add_batches()` indicates that an empty block should be created.
+    async fn try_send_batches(&self) {
```

Comment:
This will lock the `ready_batches` queue while the block is being built, right? If so, what would happen to the batches which arrive during this time? I'm assuming the batches would still get built, but nothing will be added to the `ready_batches` - and then we'll add them all at once?

Reply:
Agreed, I'll fix it tomorrow to start a new task to call `block_builder.build_block()` instead of keeping the lock on `ready_batches` for the entire duration

Reply:
Actually I had the same reasoning as mentioned [here](https://github.com/0xPolygonMiden/miden-node/pull/47#discussion_r1370077111); let's discuss in our call